File: graph/traverse/view/rightView.java
//right to left, level order
public List<TreeNode> rightView(TreeNode root){
   List<Integer> result = new ArrayList<Integer>();
   if(root == null)
    return result;

   Queue<TreeNode> queue = new LinkedList<>();
   queue.add(root);
   //On each level of the tree, we add the right-most element to the results.
   while(!queue.isEmpty()){
       //get size here
       int size = queue.size();  //each level's size 
       //each level, use a for loop but only get the first one of the each loop
       for(int i=0; i<size; i++){
           TreeNode top = queue.poll();
           //the first element in the queue (right-most of the tree)
           if(i == 0){
               result.add(top.val);
           }
           //add right first
           if(top.right != null){
               queue.offer(top.right);
           }
           //add left
           if(top.left != null){
               queue.offer(top.left);
           }
       }
   }

   return res;
}

File: tree/view/rightView.py
# iterative O(n) level order, right to left
def rightSideView(root):
    if root is None:
        return
    queue = collections.deque([root])
    res = []
    while queue:
        for index in range(len(queue)):
            top = queue.popleft()
            if index == 0:
                res.append(top.value)
            if top.right:
                queue.append(top.left)
            if top.left:
                queue.append(top.right)
    return res

def rightsideView(root):
    view = []
    if root:
        level = [root]
        while level:
            view += level[-1].value,  # for each level, only add the rightmost child
            level = [child for node in level for child in (node.left, node.right) if child]
            # for child(node.left, node.right) of all nodes on this level
    return res



# recursion: O(n)
def rightsideView(root):

#
#
# Solution 2: Recursive, first come first serve: 9 lines, 48 ms
#
# DFS-traverse the tree right-to-left, add values to the view whenever we first reach a new record depth. This is O(n).
#
# def rightSideView(self, root):
#     def collect(node, depth):
#         if node:
#             if depth == len(view):
#                 view.append(node.val)
#             collect(node.right, depth+1)
#             collect(node.left, depth+1)
#     view = []
#     collect(root, 0)
#     return view
